packages/flutter/lib/src/services/app_messages.dart

```diff
@@ -0,0 +1,64 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:flutter/shell.dart';
+import 'package:mojo/core.dart' as core;
+import 'package:sky_services/flutter/platform/app_messages.mojom.dart';
+
+/// APIs for exchanging messages with the host application.
+
+ApplicationMessagesProxy _initHostAppMessagesProxy() {
+  ApplicationMessagesProxy proxy = new ApplicationMessagesProxy.unbound();
+  shell.connectToService(null, proxy);
+  return proxy;
+}
+
+final ApplicationMessagesProxy _hostAppMessagesProxy = _initHostAppMessagesProxy();
+
+/// Send a message to the host application.
+Future<String> sendToHost(String messageName, String message) async {
+    return (await _hostAppMessagesProxy.ptr.sendString(messageName, message)).reply;
+}
+
+typedef dynamic HostMessageCallback(String message);
+typedef dynamic _MojoResponseFactory(dynamic response);
+
+class _ApplicationMessagesImpl extends ApplicationMessages {
+  final Map<String, HostMessageCallback> handlers = <String, HostMessageCallback>{};
+
+  _ApplicationMessagesImpl() {
+    shell.provideService(ApplicationMessages.serviceName,
+      (core.MojoMessagePipeEndpoint endpoint) {
+        ApplicationMessagesStub stub = new ApplicationMessagesStub.fromEndpoint(endpoint);
+        stub.impl = this;
+      }
+    );
+  }
+
+  @override
+  dynamic sendString(String messageName, String message, [_MojoResponseFactory responseFactory]) {
```

Comment:
This is so that both proxies and stubs can inherit from the same interface. In the case of a stub, it is guaranteed to be non-null.